lib/ASRouterTargeting.jsm

```diff
@@ -2,13 +2,50 @@ ChromeUtils.import("resource://gre/modules/components-utils/FilterExpressions.js
 ChromeUtils.import("resource://gre/modules/Services.jsm");
 ChromeUtils.defineModuleGetter(this, "AddonManager",
   "resource://gre/modules/AddonManager.jsm");
+ChromeUtils.defineModuleGetter(this, "NewTabUtils",
+  "resource://gre/modules/NewTabUtils.jsm");
 ChromeUtils.defineModuleGetter(this, "ProfileAge",
   "resource://gre/modules/ProfileAge.jsm");
-ChromeUtils.import("resource://gre/modules/Console.jsm");
 
 const FXA_USERNAME_PREF = "services.sync.username";
 const ONBOARDING_EXPERIMENT_PREF = "browser.newtabpage.activity-stream.asrouterOnboardingCohort";
 
+const {activityStreamProvider: asProvider} = NewTabUtils;
+
+const FRECENT_SITES_UPDATE_INTERVAL = 6 * 60 * 60 * 1000; // Six hours
+const FRECENT_SITES_IGNORE_BLOCKED = true;
+const FRECENT_SITES_NUM_ITEMS = 50;
+const FRECENT_SITES_MIN_FRECENCY = 100;
+
+const TopFrecentSitesCache = {
+  _lastUpdated: 0,
+  _topFrecentSites: null,
+  get topFrecentSites() {
+    return new Promise(async resolve => {
+      const now = new Date();
+      if (now - this._lastUpdated >= FRECENT_SITES_UPDATE_INTERVAL) {
+        this._topFrecentSites = await asProvider.getTopFrecentSites({
+          ignoreBlocked: FRECENT_SITES_IGNORE_BLOCKED,
+          numItems: FRECENT_SITES_NUM_ITEMS,
+          topsiteFrecency: FRECENT_SITES_MIN_FRECENCY,
+          onePerDomain: true,
+          includeFavicon: false
+        });
+        this._lastUpdated = now;
+      }
+      resolve(this._topFrecentSites);
+    });
+  },
+  expire() {
+    this._lastUpdated = 0;
+    this._topFrecentSites = null;
+  }
+};
+
+function getHost(url) {
+  return url.match(/:\/\/([^/]+)/)[1];
```

Comment:
can you use the `URL` constructor here / the hostname property instead of a regex